rest/oidc_api_test.go

```diff
@@ -1,139 +1,612 @@
-		validationKeyGoogle     = "validation.key.google"
-	)
-
-	// Mock OpenID Connect Provider Discovery Endpoints
-	var issuerGoogle, issuerSalesforce string
-	mux := http.NewServeMux()
-	mux.HandleFunc("/google"+wellKnownPath, func(res http.ResponseWriter, req *http.Request) {
-		res.Header().Set("Content-Type", "application/json")
-		body := `{
-			"issuer": "${issuer}", 
-			"authorization_endpoint": "https://accounts.google.com/o/oauth2/v2/auth", 
-        	"token_endpoint": "https://oauth2.googleapis.com/token",
-			"jwks_uri": "https://www.googleapis.com/oauth2/v3/certs", 
-			"id_token_signing_alg_values_supported": ["RS256"]
-		}`
-		_, err := io.WriteString(res, strings.ReplaceAll(body, "${issuer}", issuerGoogle))
-		require.NoError(t, err, "Failed to write response body for google provider")
-	})
-	mux.HandleFunc("/salesforce"+wellKnownPath, func(res http.ResponseWriter, req *http.Request) {
-		res.Header().Set("Content-Type", "application/json")
-		body := `{
-			"issuer": "${issuer}", 
-			"authorization_endpoint": "https://login.salesforce.com/services/oauth2/authorize",
-        	"token_endpoint": "https://login.salesforce.com/services/oauth2/token",
-			"jwks_uri": "https://login.salesforce.com/id/keys", 
-			"id_token_signing_alg_values_supported": ["RS256"]
-		}`
-		_, err := io.WriteString(res, strings.ReplaceAll(body, "${issuer}", issuerSalesforce))
-		require.NoError(t, err, "Failed to write response body for salesforce provider")
+// mockOptions keeps options required for mocking
+// auth code flow responses.
+type mockOptions map[string]interface{}
+
+var (
+	optionsOnce sync.Once
+	options     mockOptions
+)
+
+func NewMockOptions() mockOptions {
+	optionsOnce.Do(func() {
+		options = make(mockOptions)
 	})
+	return options
+}
+
+func (options mockOptions) Clear() {
+	for option := range options {
+		delete(options, option)
+	}
+}
+
+// mockAuthServer mocks a fake OAuth server by registering mock handlers
+func mockAuthServer() (*httptest.Server, error) {
+	router := mux.NewRouter()
+	router.HandleFunc("/{provider}"+auth.DiscoveryConfigPath, mockDiscoveryHandler).Methods(http.MethodGet)
+	router.HandleFunc("/google/auth", mockAuthHandler).Methods(http.MethodGet, http.MethodPost)
+	router.HandleFunc("/google/token", mockTokenHandler).Methods(http.MethodPost)
+	return httptest.NewServer(router), nil
+}
+
+// mockDiscoveryHandler mocks the provider discovery endpoint with a mock response
+func mockDiscoveryHandler(w http.ResponseWriter, r *http.Request) {
+	var metadata auth.ProviderMetadata
+	vars := mux.Vars(r)
+	provider := vars["provider"]
+	if provider == "google" {
+		issuerGoogle := options["issuerGoogle"].(string)
+		metadata = auth.ProviderMetadata{
+			Issuer: issuerGoogle, AuthorizationEndpoint: issuerGoogle + "/auth",
+			TokenEndpoint: issuerGoogle + "/token", JwksUri: issuerGoogle + "/oauth2/v3/certs",
+			IdTokenSigningAlgValuesSupported: []string{"RS256"},
+		}
+	} else if provider == "facebook" {
```

Comment:
If these are just test issuers/providers, is there a good reason to name them 'google' and 'facebook'? Are these testing configurations or token formats specific to those real-world providers? If not, keeping the tests generic would be preferable, IMO.

Reply:
Yeah, these are just test issuers and providers and there is no good reason to name them 'google' and 'facebook'; just given this name. These names came to my thought at the time of writing this code and the token formats NOT specific to those real-world providers. I see what you are suggesting, added the changes to make these tests more generic.